### ISSUE TITLE
Change `report` to be consistent and predictable in how it handles the provided position arguments

### ISSUE THREAD

**Issue by romainmenke** [stylelint/stylelint#8126]
This same issue is what I was also trying to address in : https://github.com/stylelint/stylelint/pull/8098

Currently I see these behaviors:

1. if `line` is provided:
   - use `line` as `startLine`
2. if `line` is not provided:
   - calculate `startLine` from `index` and `endIndex` (ignoring `start.line` if that was provided)
   - if `index` was not provided, PostCSS will fallback to the start of the node
3. use `startLine` to determine disable ranges
3. `startLine` is no longer used after this point
4. if `start` was provided
   - use `start` in `warningProperties`
5. if `index` was provided and is a `number`
   - use `index` in `warningProperties`
6. if `end` was provided
   - use `end` in `warningProperties`
7. if `endIndex` was provided and is a `number`
   - use `endIndex` in `warningProperties`
8. if none were provided, let PostCSS fallback to the entire range of the node

----------

This implies several inconsistencies:
1. passing only `start` and `end` will not affect disable ranges
2. passing only `line` will not be reflected in the report, even when it is used in the disable ranges

Possibly there are more bugs but I think these are the most important to resolve.

------

Issue 1 is also why it is still needed to pass a `line` argument in: https://github.com/stylelint/stylelint/pull/8074/files#diff-bd94fe99a5bd46c7e6b11e8afd5f71ca4aa1897e2d52801cc1e7e3be5de7d925R117

Changing that rule report to pass `index` and `endIndex` instead makes the `line` argument redundant.

This is very counter intuitive from a rule author perspective because currently `line` is already passed implicitly as part of `start` and yet it seems to be ignored.

-----

My preferred solution would be to have proper input validation and normalization for the position arguments of `report()`
The resulting values can then be used in a more consistent and controlled manner.

**Comment by Mouvedia** [comment]
The only thing that bothers me with `report` is that `node` is/was treated as if it is optional.

https://github.com/stylelint/stylelint/blob/3bf048d2e0c7d67841e0fcc7298788ae89c5a7e9/lib/utils/report.mjs#L56

https://github.com/stylelint/stylelint/blob/3bf048d2e0c7d67841e0fcc7298788ae89c5a7e9/lib/utils/report.mjs#L52

https://github.com/stylelint/stylelint/blob/3bf048d2e0c7d67841e0fcc7298788ae89c5a7e9/lib/utils/report.mjs#L89

I reckon it should have been settled a long time ago but there's probably code in the wild relying on that unclear state.
Before overhauling it completely, shouldn't we resolve that unknown first?
i.e. either `node?: PostCSS.Node` or refactor `report`

---

You will probably consider this off-topic but I think it would be kind of weird to start a refactor without either fixing the type of `Problem` or explaining—with comment(s)—why we have so much defensive programming in place for `node`.

**Comment by Mouvedia** [comment]
> what combinations of inputs are allowed

Alright for once I would prefer if we were to propose the types _before_ opening the PR.
e.g.
```ts
type Foo =
	| {  }
	| {  };

export type Problem = {
  // common properties
} & Foo;
```

> We can express in the types which combinations of types make sense and will produce good results.

👍 

If the change makes perfect sense then it can be part of the next minor.

> I agree that it would be simpler without those checks, but it didn't really bother me :)

The `node` issue is just bothering me, because I emphasizes types way more in a non-typescript codebase.
It's probably over-zealous of me.

**Comment by romainmenke** [comment]
Is it useful to be able to set partial positions?
Only `index` or `endIndex` and only `start` or `end`?

Current behaviors:
- passing `index` and omitting `endIndex`:
  - if `index` is zero, set `endIndex` to the end of the node
  - if `index` is non-zero, set `endIndex` to `index + 1` ([see](https://github.com/postcss/postcss/blob/aed8b899884db4bc89f60782771eecd8c7861b32/lib/node.js#L290-L292), not sure if this is a bug or a feature)
- passing `endIndex` and omitting `index`:
  - set `index` to the start of the node
- passing `start` and omitting `end`:
  - set `end` to `start.column + 1` ([see](https://github.com/postcss/postcss/blob/aed8b899884db4bc89f60782771eecd8c7861b32/lib/node.js#L256-L259)
- passing `end` and omitting `start`:
  - set `start` to the start of the node

-----

```ts
export type PositionArguments = {
		start: Position;
		end?: Position;
	} | {
		start?: Position;
		end: Position;
	} | {
		index: number;
		endIndex?: number;
	} | {
		index?: number;
		endIndex: number;
	} | {
		line: number;
	} | {
		word: string;
	};
```

vs.

```ts
export type PositionArguments = {
		start: Position;
		end: Position;
	} | {
		index: number;
		endIndex: number;
	} | {
		line: number;
	} | {
		word: string;
	};
```

With maybe only the `end` and `endIndex` positions as optional arguments?

```ts
export type PositionArguments = {
		start: Position;
		end?: Position;
	} | {
		index: number;
		endIndex?: number;
	} | {
		line: number;
	} | {
		word: string;
	};
```

**Comment by Mouvedia** [comment]
> if `index` is non-zero, set `endIndex` to `index + 1`
> […]
> set end to start.column + 1

I would have expected `endIndex`/`range.end.column` to either be:

- end of the node
- end of the line (e.g. detect first newline)

Based on that I would pick:

```
export type PositionArguments = {
		start?: Position;
		end: Position;
	} | {
		index: 0;
		endIndex?: number;
	} | {
		index?: number;
		endIndex: number;
	} | {
		line: number;
	} | {
		word: string;
	};
```

That would be a breaking change though.

---

Maybe reflect the status quo for now with a refactor and create a separate issue if you feel a breaking change is in order?

**Comment by romainmenke** [comment]
> Such changes imply a semver major and ideally we can already make some improvements in a patch or minor.
But we can decide a general direction where we want to go and split out into multiple issues, some of which might be postponed until a next major.

**Comment by Mouvedia** [comment]
> `+ 1`
> `not sure if this is a bug or a feature`

It's a feature when you target the end of something.
e.g. `hue-degree-notation` for option `"number"`

see also: https://github.com/stylelint/stylelint/issues/7192#issuecomment-2338068550

It's a source of bugs if you target one character that may be at the start of the line.
e.g. a rule that checks if something is capitalized would easily trip a plugin author if he doesn't pass an `endIndex`
```css
a {
  property:
value;
}
```

> Lets first decide on a direction and then see how we can get there.

👍

**Comment by romainmenke** [comment]
I think the approach that is least ambiguous and has the fewest surprises for rule authors would be:
- `node` is always required
- only provide position information through one means at a time (either `word`, `line`, `index/endIndex` or `start/end`)
- `index/endIndex` must always both be present if either is present
- `start/end` must always both be present if either is present
- `line` is still allowed but we calculate a range from it, covering that entire line
- `word` is kept as it exists today

And from this hypothetical situation we can maybe relax parts if we think it is really useful to rule authors?

-----

I think we have been changing two aspects in a good way and in the right direction:
- enabling configuration comments inside selectors, values, at rules, ... 
- better fixers and allowing fixers when configuration comments are present

But that Stylelint and reports now need to be a lot more accurate with positions.
Stylelint was more resistant to bugs because it was less capable before we added those two improvements.

To me it makes sense to trade same freedom/flexibility when passing position arguments for more accuracy and control.
So that the whole works better.

But also important to consciously pick the parts where flexibility is still useful.
e.g. maybe there really is a strong use case for passing only `endIndex`?

I would like to hear the thoughts of others on this, if specific cases that would eventually become disallowed should actually be preserved. Not because they used to work, but because they really solve a specific problem rule authors face.

-----

[off topic]

Isn't a bug that PostCSS does `column + 1` when `end` is omitted?
What if `start` was the last char on that line, then `end` will point to a column that doesn't exist, right?

I haven't tested this, so I don't know how this surfaces.

**Comment by ybiquitous** [comment]
@romainmenke Thank you for opening the discussion. I totally agree with your thoughts. We should fix this problem in the next major version through deprecation minor releases. 👍🏼 

Also, I am in favor of your suggestion below. This makes sense to me. 😃 
  
> * `node` is always required
> * only provide position information through one means at a time (either `word`, `line`, `index/endIndex` or `start/end`)
> * `index/endIndex` must always both be present if either is present
> * `start/end` must always both be present if either is present
> * `line` is still allowed but we calculate a range from it, covering that entire line
> * `word` is kept as it exists today

If you're okay too, as @Mouvedia did, can you express this new restriction with the TypeScript language to make it clear?

**Comment by Mouvedia** [comment]
> * `node` is always required
> * only provide position information through one means at a time (either `word`, `line`, `index/endIndex` or `start/end`)

probably non-breaking 
  
> * `index/endIndex` must always both be present if either is present
> * `start/end` must always both be present if either is present

breaking changes

> * `line` is still allowed but we calculate a range from it, covering that entire line
> * `word` is kept as it exists today

non-breaking
 
> And from this hypothetical situation we can maybe relax parts if we think it is really useful to rule authors?

👍 

> I would like to hear the thoughts of others on this […]

👍

**Comment by romainmenke** [comment]
Proposed types:

```ts
export type ProblemLocation = {
	/**
	 * The inclusive start index of the problem, relative to the node's
	 * source text.
	 */
	index: number;
	/**
	 * The exclusive end index of the problem, relative to the node's
	 * source text.
	 */
	endIndex: number;
} | {
	/**
	 * The inclusive start position of the problem, relative to the
	 * node's source text. If provided, this will be used instead of
	 * `index`.
	 */
	start: Position;
	/**
	 * The exclusive end position of the problem, relative to the
	 * node's source text. If provided, this will be used instead of
	 * `endIndex`.
	 */
	end: Position;
} | {
	word: string;
} | {
	line: number;
}

/**
 * A lint problem.
 */
export type Problem = {
	ruleName: string;
	result: PostcssResult;
	message: RuleMessage;
	messageArgs?: Parameters<RuleMessageFunc> | undefined;
	node: PostCSS.Node;
	/**
	 * The inclusive start index of the problem, relative to the node's
	 * source text.
	 */
	index?: number;
	/**
	 * The exclusive end index of the problem, relative to the node's
	 * source text.
	 */
	endIndex?: number;
	/**
	 * The inclusive start position of the problem, relative to the
	 * node's source text. If provided, this will be used instead of
	 * `index`.
	 */
	start?: Position;
	/**
	 * The exclusive end position of the problem, relative to the
	 * node's source text. If provided, this will be used instead of
	 * `endIndex`.
	 */
	end?: Position;
	word?: string;
	line?: number;
	/**
	 * Optional severity override for the problem.
	 */
	severity?: RuleSeverity;
	fix?: () => void | undefined | never;
} & ProblemLocation;
```


A possible implementation of extracting range information from a report:

```js
/** @param {Problem} problem */
function problemRange(problem) {
	// `start/end` based range
	if (problem.start && problem.end) {
		return {
			start: problem.start,
			end: problem.end
		}
	}

	// `index/endIndex` based range
	if (isNumber(problem.index) && isNumber(problem.endIndex)) {
		return problem.node.rangeBy({ index: problem.index, endIndex: problem.endIndex })
	}
	
	// `word` based range
	if (problem.word) return problem.node.rangeBy({ word: problem.word })
	
	// `line` based range
	if (isNumber(problem.line)) {
		// Calculate the range of the given line

		const start = {
			line: problem.line,
			column: 1,
		};

		const end = {
			line: problem.line,
			column: 1,
		}

		const inputCSS = problem.node.source?.input?.css ?? '';

		let column = 1;
		let line = 1;
		let startOffset = 0;

		if (isNumber(problem.node.source?.start?.offset)) {
			startOffset = problem.node.source?.start?.offset;
		}

		for (let i = startOffset; i < inputCSS.length; i++) {
			if (inputCSS[i] === '\n') {
				if (line === end.line) {
					end.column = column

					break;
				}

				column = 1;
				line += 1;
			} else {
				column += 1;
			}
		}

		return {
			start,
			end
		}
	}

	// Insufficient or incorrect arguments
	throw new Error(
		// TODO: write a clear and correct message
		`The "${problem.ruleName}" rule failed to pass correct position arguments to the \`report()\` function.`,
	);
}
```

There is still quite some complexity to convert a `line` argument to a proper range.
I think this is fine, but we could also consider dropping `line` entirely.

**Comment by romainmenke** [comment]
> `line` is still allowed but we calculate a range from it, covering that entire line

We could also limit this to the first column of the given line.
This would dramatically simplify implementation.

Since `line` isn't that useful or accurate I think this might be acceptable.

**Comment by ybiquitous** [comment]
Thanks for providing the type. That sounds good to me 👍🏼 

I have the same concern about `line`. If we can give it up, I'd like to deprecate `line` and keep things simple.

**Comment by romainmenke** [comment]
I think it should also be supported to omit all location arguments and only provide `node`.
This would implicitly expand to the `start` and `end` of the given `node`. 

Then there is still an easy path for rules that want to indicate an error somewhere in a node without utilizing specialized parsers or more involved logic. `word` could be used to the same effect but this has more side effects.

**Comment by ybiquitous** [comment]
It makes sense to require only a node argument and make others optional 👍

Now I think this is ready to implement if you have no concerns.😀

**Comment by romainmenke** [comment]
Indeed.

In 16.x we can do these things:
- add deprecation warnings
- update core rules so that these are ready for the upcoming breaking changes

**Comment by romainmenke** [comment]
I've prepared a draft implementation of the planned changes in https://github.com/stylelint/stylelint/pull/8217

This is intended as a guide for the changes to rules and the deprecations.
I've also patched the core rules and found that the needed changes were fairly minimal.

For now the general types and validation of those types is important.
Please give feedback or suggestions to improve this 🙇 

If this looks fine we can proceed with the deprecations.

**Comment by Mouvedia** [comment]
> If this looks fine we can proceed with the deprecations.

LGTM

1. `| {}` is quite permissive.
2. I think that adding
```
| {
	index: 0;
	endIndex?: number;
}
```
would be warranted.

**Comment by romainmenke** [comment]
> ```ts
> | {
> 	index: 0;
> 	endIndex?: number;
> }
> ```
> should be added to `ProblemLocation`.

I don't think this allows for novel capabilities or provides a sufficiently useful shortcut to warrant the potential ambiguity. It will always be unclear if omitting an `end` marker implies a range of one column or if it implies to take the end of the current PostCSS node.

Was there a specific case or reason for wanting this @Mouvedia?

We can more easily relax the typings in a minor/patch than we can make the types stricter.
So I currently favor starting out with a solid but minimal set of valid types and gather feedback from there.

But maybe I am overlooking a very common use case?

**Comment by romainmenke** [comment]
> `| {}` is quite permissive.

It is! I wasn't sure how to express an optional union of types.
The `Problem` type can be a union with any of these or, optionally none at all.

Adding `| {}` seemed to work, even if very permissive.

Is there a more idiomatic way of writing this type?

**Comment by Mouvedia** [comment]
> Is there a more idiomatic way of writing this type?

This won't answer your question but you might find this note from the [`@typescript-eslint/no-empty-object-type` rule](https://typescript-eslint.io/rules/no-empty-object-type/) interesting.

> Note that this rule does not report on:
>
> - {} as a type constituent in an intersection type (e.g. types like TypeScript's built-in type NonNullable<T> = T & {}), as this can be useful in type system operations.
> - […]

Maybe you want `| Record<string, never>`, I am not sure.

**Comment by Mouvedia** [comment]
> We can more easily relax the typings in a minor/patch than we can make the types stricter.

Adding that type would make the union less strict, not stricter, because

```
		| {
				index: number;
				endIndex: number;
		  }
```

doesn't allow to not pass `endIndex`.
Of course what I just said doesn't account for `| {}`.

> Was there a specific case or reason for wanting this @Mouvedia?

I think that the case

> [if `index` is zero, set `endIndex` to the end of the node](https://github.com/stylelint/stylelint/issues/8126#issuecomment-2478226329)

is pretty common and useful.

**Comment by ybiquitous** [comment]
Omitting `endIndex` seems pretty ambiguous to me, too. I think specifying both the start and end is much more obvious.

**Comment by romainmenke** [comment]
> > We can more easily relax the typings in a minor/patch than we can make the types stricter.

> Adding that type would make the union less strict, not stricter

Exactly :)

We can start with the types I've proposed in for example 17.0 as that change would be breaking.
But if there is a lot of feedback we can add your proposed addition in 17.1 as it wouldn't be breaking.

But we can't do the inverse.

We can't ship your proposal in 17.0, find that it confuses people, and then patch it out. Removing it would need to wait for 18.0

Starting with the strictest type in a next major gives us more options to adjust later in a minor/patch.

-------

> I think that the case
> > if index is zero, set endIndex to the end of the node
> is pretty common and useful.

Instead of passing `{ index: 0 }`, rule authors shouldn't pass anything at all.
When omitting all range related arguments the fallback will be the entire node range.
This is the same as setting `{ index: 0 }` in the current version of Stylelint/PostCSS but less ambiguous and less code.

**Comment by Mouvedia** [comment]
> Omitting `endIndex` seems pretty ambiguous to me, too. I think specifying both the start and end is much more obvious.

Welp what he proposed is fine to me because `| {}` would also permit it.
I was just trying to be more granular and precise, in particular if we were to replace `| {}` later on.

**Comment by Mouvedia** [comment]
As explained in https://github.com/stylelint/stylelint/issues/8126#issuecomment-2483032339, I think we should have
```
| {
	index: number;
	endIndex?: number;
}
```
instead of
```
| {
	index: number;
	endIndex: number;
}
```

The reason being that there are too many plugin rules in the wild that would need to be fixed.
It should output a deprecation warning though.
@romainmenke You can check [these results](https://github.com/search?q=stylelint+%22report%28%22+%22index%3A%22+OR+%22index%2C%22+NOT+%22endIndex%3A%22+NOT+%22endIndex%2C%22+language%3ATypeScript+OR+language%3AJavaScript&type=code) to give you an approximate idea of the number of plugins that it would affect.

**Comment by romainmenke** [comment]
Is there any reason to have `endIndex?` going forwards, other than back compat?

Having `endIndex` values will also give a better experience for end users as the report locations will better correspond with the problematic CSS source.

In the short term we will cause inconvenience to rule authors but we make the API less ambiguous and we improve the overall experience for CSS authors.

-----

It is always implied that we adjust plans if we receive feedback.
A deprecation cycle is a two way communication, we inform downstream users of an upcoming change and downstream users can inform us that a change is going to fast or is removing an important feature.

We also do deprecation cycles exactly because we know a feature is in use.

Possibly rule authors find that adding `endIndex` is low effort and the ecosystem quickly migrates to this new type? The only way to learn more is to do the deprecation cycle.

I do hear and understand your concerns but I think this change might be minor and that the overal benefits weigh up to the expected effort.

**Comment by Mouvedia** [comment]
Would not supporting
```
| {
	index: 0;
	endIndex?: number;
}
```
specifically and supporting
```
| {
	index: number;
	endIndex?: number;
}
```
for a deprecation cycle be satisfactory?
i.e. accompanied by a deprecation warning

If that's what you were planning from the start, Ill apologise for the confusion.

**Comment by romainmenke** [comment]
I am not sure I follow :)

For the deprecation cycle no types would be changed, because changing the types would already be breaking.

We need to add deprecation warnings first and only do the type changes in the next major.

The proposed draft is only to illustrate the state we aim for in the next major.

**Comment by Mouvedia** [comment]
I was using types to illustrate my points.
We were agreeing all along; good, let's move on then.

**Comment by romainmenke** [comment]
Thank you for the feedback! Happy to have found consensus :)

I've split out the deprecations and needed changes to core rules into sub issues as work can already start on those.

Going forwards this issue will focus on the eventual breaking changes planned for the next major.

**Comment by jeddy3** [comment]
Closed by https://github.com/stylelint/stylelint/pull/8217
